peer/brontide.go

```diff
@@ -957,18 +960,29 @@ func (p *Brontide) readNextMessage() (lnwire.Message, error) {
 			return readErr
 		}
 
-		rawMsg, readErr = noiseConn.ReadNextBody(buf[:pktLen])
-		return readErr
-	})
-	atomic.AddUint64(&p.bytesReceived, uint64(len(rawMsg)))
-	if err != nil {
-		return nil, err
-	}
+		// The ReadNextBody method will actually end up re-using the
+		// buffer, so within this closure, we can continue to use
+		// rawMsg as it's just a slice into the buf from the buffer
+		// pool.
+		rawMsg, readErr := noiseConn.ReadNextBody(buf[:pktLen])
```

Comment:
So we allocate once from reader pool, hence the `buf`, then `rawMsg` also is encoded in that `buf`, then `lnwire.ReadMessage` is too using that `buf`?

Reply:
Yeh exactly! That's the idea at least, it may be possible to dip down a level to truly confirm this by looking at the heap allocation of a sample program using `pprof`. 

Elaborating a bit, before this set of changes, `ReadNextBody` would use the passed buffer to read out the _encrypted_ message contents into a local buffer. Then allocate a _new_ buffer to store the decrypted contents. Now it will just re-use the passed in `buf` to store the decrypted contents in line. So `buf` and `rawMsg` point to the exact same byte slice.

Reply:
I think this is a very cool usage of the slice. As the slice never grows but only re-sliced, the pointer never changed and we could always reuse that piece of memory!